After a reconnect-scheduled, if the login times out, we never connect again

```
2017-03-21 20:51:12.008 [ INFO] - 1624938ms	satinnape_Ey8B9VpKM (happnerEndpoint) happner-endpoint reconnect/scheduled
2017-03-21 20:51:38.473 [ERROR] - 26468ms	client (HappnClient) unhandled error Error: api request timed out action: login
    at Object.<anonymous> (/node_js/app/node_modules/happn/lib/client/base.js:691:29)
    at ontimeout (timers.js:365:14)
    at tryOnTimeout (timers.js:237:5)
    at Timer.listOnTimeout (timers.js:207:5)
2017-03-21 20:51:38.482 [ERROR] - Error: api request timed out action: login
    at Object.<anonymous> (/node_js/app/node_modules/happn/lib/client/base.js:691:29)
    at ontimeout (timers.js:365:14)
    at tryOnTimeout (timers.js:237:5)
    at Timer.listOnTimeout (timers.js:207:5)
```

--- a/RELEASES.md
+++ b/RELEASES.md
@@ -409,3 +409,8 @@
 -----------------
 
 - Issue 152 - Destroy pubsub if login fails so that we don't have primus connections retrying
+
+2.17.2 2017-02-08
+-----------------
+
+- Issue 157 - Retry login if it timed out after a reconnect.

--- a/lib/client/base.js
+++ b/lib/client/base.js
@@ -518,10 +518,7 @@
     });
 
     function handleLoginCallback(e) {
-      if (e && !_this.initialized) { // this is the first login attempt
-        if (_this.pubsub && _this.pubsub.destroy) _this.pubsub.destroy();
-        _this.pubsub = null;
-      }
+      // JW - Reversing this change. This should be handled by the user of this client.
       callback(e);
     }
 
@@ -631,6 +628,7 @@
   }
 
   HappnClient.prototype.handle_reconnect_scheduled = function (opts) {
+    this._reconnectSuccessful = false;
     this.emit('reconnect-scheduled', opts)
   }
 
@@ -745,11 +743,9 @@
   HappnClient.prototype.reconnect = function (options) {
     var _this = this;
 
-    if (!this.reconnectAttempt) this.reconnectAttempt = 0;
-    var reconnectAttempt = ++this.reconnectAttempt;
-
     _this.authenticate(function (e) {
-      if (_this.reconnectAttempt != reconnectAttempt) return; // this is not the last attempt, ignore it
+
+      if (_this._reconnectSuccessful) return;
 
       if (e) {
         if (e.message && e.message.indexOf('api request timed out action') == 0) return _this.reconnect();
@@ -764,6 +760,7 @@
         });
       });
 
+      _this._reconnectSuccessful = true;
       _this.emit('reconnect-successful', options)
     });
   }

--- a/lib/client/base.js
+++ b/lib/client/base.js
@@ -518,7 +518,7 @@
     });
 
     function handleLoginCallback(e) {
-      if (e) {
+      if (e && !_this.initialized) { // this is the first login attempt
         if (_this.pubsub && _this.pubsub.destroy) _this.pubsub.destroy();
         _this.pubsub = null;
       }
@@ -533,12 +533,12 @@
       return Primus.connect(this.options.config.url, this.options.config.pubsub.options);
     }
     else {
-      Primus = require('primus'),
-        Socket = Primus.createSocket({
-          transformer: this.options.config.transformer,
-          parser: this.options.config.parser,
-          manual: true
-        });
+      Primus = require('primus');
+      Socket = Primus.createSocket({
+        transformer: this.options.config.transformer,
+        parser: this.options.config.parser,
+        manual: true
+      });
       return new Socket(this.options.config.url, this.options.config.pubsub.options);
     }
   };
@@ -745,9 +745,16 @@
   HappnClient.prototype.reconnect = function (options) {
     var _this = this;
 
-    _this.authenticate(function (e) {
+    if (!this.reconnectAttempt) this.reconnectAttempt = 0;
+    var reconnectAttempt = ++this.reconnectAttempt;
 
-      if (e) return _this.handle_error(e, 3);
+    _this.authenticate(function (e) {
+      if (_this.reconnectAttempt != reconnectAttempt) return; // this is not the last attempt, ignore it
+
+      if (e) {
+        if (e.message && e.message.indexOf('api request timed out action') == 0) return _this.reconnect();
+        return _this.handle_error(e, 3);
+      }
 
       Object.keys(_this.events).forEach(function (eventPath) {
         var listeners = _this.events[eventPath];

--- a/lib/client/base.js
+++ b/lib/client/base.js
@@ -743,6 +743,10 @@
   HappnClient.prototype.reconnect = function (options) {
     var _this = this;
 
+    if (!_this.initialized) return;
+
+    this.emit('reconnect');
+
     _this.authenticate(function (e) {
 
       if (_this._reconnectSuccessful) return;

--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "happn",
   "description": "pub/sub api as a service using primus and mongo & redis or nedb, can work as cluster, single process or embedded using nedb, use in production at your own risk",
-  "version": "2.17.1",
+  "version": "2.17.2",
   "main": "./lib/index",
   "protocol":"1.1.0",
   "scripts": {

--- a/test/c5_client_events.js
+++ b/test/c5_client_events.js
@@ -118,36 +118,6 @@ describe('c5_client_events', function () {
 
   });
 
-  it('does not retry after a login has failed', function (callback) {
-    this.timeout(100000);
-
-    var eventsFired = {
-      'reconnect-scheduled': false,
-      'reconnect-successful': false
-    };
-
-    var client = (new happn.client()).client({
-      config: {
-        username: '_ADMIN',
-        password: 'bad password',
-        port: 55002
-      }
-    });
-
-    client.initialize(function (e) {
-      if (e && e.code == 'ECONNREFUSED') return;
-      eventsFired['reconnect-scheduled'] = false;
-      setTimeout(function () {
-        expect(eventsFired['reconnect-scheduled']).to.eql(false);
-        callback();
-      }, 5000);
-    });
-
-    client.onEvent('reconnect-scheduled', function (opts) {
-      eventsFired['reconnect-scheduled'] = true;
-    });
-  });
-
   require('benchmarket').stop();
 
 });

--- a/test/e7-login-failures-reconnect.js
+++ b/test/e7-login-failures-reconnect.js
@@ -4,7 +4,7 @@ var Promise = require('bluebird');
 
 var service1Name;
 //checks info is stored next to login
-describe('e5-login-digest', function () {
+describe(require('path').basename(__filename), function () {
 
   this.timeout(60000);
 

--- a/test/e7-login-failures-reconnect.js
+++ b/test/e7-login-failures-reconnect.js
@@ -1,5 +1,6 @@
 var Happn = require('../');
 var expect = require('expect.js');
+var Promise = require('bluebird');
 
 var service1Name;
 //checks info is stored next to login

--- a/test/e7-login-failures-reconnect.js
+++ b/test/e7-login-failures-reconnect.js
@@ -1,0 +1,159 @@
+var Happn = require('../');
+var expect = require('expect.js');
+
+var service1Name;
+//checks info is stored next to login
+describe('e5-login-digest', function () {
+
+  this.timeout(60000);
+
+  var server1;
+
+  var testGroup2 = {
+    name: 'TEST GROUP2',
+    custom_data: {
+      customString: 'custom2',
+      customNumber: 0
+    }
+  };
+
+  testGroup2.permissions = {
+    '/@HTTP/secure/test/removed/user': {actions: ['get']},
+    '/@HTTP/secure/test/not_removed/user': {actions: ['get']}
+  };
+
+  var testUser2 = {
+    username: 'TEST USER2@blah.com',
+    password: 'TEST PWD'
+  };
+
+  function linkUser() {
+    return new Promise(function (resolve, reject) {
+      var addedTestGroup2;
+      var addedTestuser2;
+
+      server1.services.security.upsertGroup(testGroup2, {overwrite: false}, function (e, result) {
+
+        if (e) return reject(e);
+        addedTestGroup2 = result;
+
+        server1.services.security.upsertUser(testUser2, {overwrite: false}, function (e, result) {
+
+          if (e) return reject(e);
+          addedTestuser2 = result;
+
+          server1.services.security.linkGroup(addedTestGroup2, addedTestuser2, function (e) {
+
+            if (e) return reject(e);
+            resolve();
+          });
+        });
+      });
+    });
+  }
+
+  function createService(allowLogin) {
+    return Happn.service.create({secure: true})
+      .then(function (server) {
+        server1 = server;
+        service1Name = server.name;
+        if (!allowLogin) {
+          server1.services.pubsub.login = function () {
+            console.log("login attempt");
+            this.emit('loginAttempt');
+          }
+        }
+      })
+      .then(linkUser);
+  }
+
+  function stopService() {
+    return new Promise(function (resolve) {
+      if (!server1) resolve();
+      server1.stop(function (e) {
+        if (e) {
+          console.warn('failed to stop server1: ' + e.toString());
+          reject(e);
+          return;
+        }
+        resolve();
+      });
+    });
+  }
+
+  before('starts the services', function () {
+    return createService(true);
+  });
+
+  after('stops the services', function () {
+    return stopService();
+  });
+
+  it('logs in normally', function () {
+    Happn.client.create(
+      {
+        config: {
+          username: testUser2.username,
+          password: testUser2.password
+        }
+      })
+      .then(function (clientInstance) {
+        return clientInstance.disconnect();
+      })
+  });
+
+  it('can login again if a login attempt failed on reconnect', function () {
+    var client;
+
+    return Happn.client.create(
+      {
+        config: {
+          username: testUser2.username,
+          password: testUser2.password
+        }
+      })
+      .then(function (clientInstance) {
+        client = clientInstance;
+        return stopService();
+      })
+      .then(function waitForDisconnect() {
+        return new Promise(function (resolve) {
+          {
+            var subHandle = client.onEvent('reconnect-scheduled',
+              function waitForReconnectScheduled() {
+                client.offEvent(subHandle);
+                resolve();
+              });
+          }
+        });
+      })
+      .then(function startServiceWithNoLogin() {
+        return createService(false);
+      })
+      .then(function waitForLoginAttempt() {
+        return new Promise(function (resolve) {
+          server1.services.pubsub.on('loginAttempt', function waitForAttempt() {
+            resolve();
+          })
+        });
+      })
+      // .then(stopService)
+      .then(function(){
+        delete server1.services.pubsub.login;
+      })
+      .then(function() {
+        return new Promise(function (resolve) {
+          {
+            var subHandle = client.onEvent('reconnect-successful',
+              function waitForConnectSuccess() {
+                client.offEvent(subHandle);
+                resolve();
+              });
+          }
+        });
+      })
+      .then(function () {
+        return client.disconnect();
+      });
+  });
+});

--- a/test/e7-login-failures-reconnect.js
+++ b/test/e7-login-failures-reconnect.js
@@ -59,7 +59,6 @@ describe('e5-login-digest', function () {
         service1Name = server.name;
         if (!allowLogin) {
           server1.services.pubsub.login = function () {
-            console.log("login attempt");
             this.emit('loginAttempt');
           }
         }
@@ -137,7 +136,6 @@ describe('e5-login-digest', function () {
           })
         });
       })
-      // .then(stopService)
       .then(function(){
         delete server1.services.pubsub.login;
       })
@@ -149,6 +147,67 @@ describe('e5-login-digest', function () {
                 client.offEvent(subHandle);
                 resolve();
               });
+          }
+        });
+      })
+      .then(function () {
+        return client.disconnect();
+      })
+  });
+
+  it('only has one reconnect-successful event', function () {
+    var client;
+
+    return Happn.client.create(
+      {
+        config: {
+          username: testUser2.username,
+          password: testUser2.password
+        }
+      })
+      .then(function (clientInstance) {
+        client = clientInstance;
+        return stopService();
+      })
+      .then(function waitForDisconnect() {
+        return new Promise(function (resolve) {
+          {
+            var subHandle = client.onEvent('reconnect-scheduled',
+              function waitForReconnectScheduled() {
+                client.offEvent(subHandle);
+                resolve();
+              });
+          }
+        });
+      })
+      .then(function startServiceWithNoLogin() {
+        return createService(false);
+      })
+      .then(function waitForLoginAttempt() {
+        return new Promise(function (resolve) {
+          server1.services.pubsub.on('loginAttempt', function waitForAttempt() {
+            resolve();
+          })
+        });
+      })
+      .then(stopService)
+      .then(function() {
+        return createService(true);
+      })
+      .then(function() {
+        return new Promise(function (resolve) {
+          {
+            var eventCalled = 0;
+            var subHandle = client.onEvent('reconnect-successful',
+              function waitForConnectSuccess() {
+                eventCalled++;
+              });
+
+            setTimeout(function() {
+              expect(eventCalled).to.equal(1);
+              client.offEvent(subHandle);
+              resolve();
+            },35000)
           }
         });
       })

--- a/test/e7-login-failures-reconnect.js
+++ b/test/e7-login-failures-reconnect.js
@@ -137,10 +137,10 @@ describe('e5-login-digest', function () {
           })
         });
       })
-      .then(function(){
+      .then(function () {
         delete server1.services.pubsub.login;
       })
-      .then(function() {
+      .then(function () {
         return new Promise(function (resolve) {
           {
             var subHandle = client.onEvent('reconnect-successful',
@@ -156,7 +156,7 @@ describe('e5-login-digest', function () {
       })
   });
 
-  it('only has one reconnect-successful event', function () {
+  it('emits when it retries to reconnect', function () {
     var client;
 
     return Happn.client.create(
@@ -192,28 +192,129 @@ describe('e5-login-digest', function () {
         });
       })
       .then(stopService)
-      .then(function() {
+      .then(function () {
         return createService(true);
       })
-      .then(function() {
+      .then(function () {
         return new Promise(function (resolve) {
           {
-            var eventCalled = 0;
-            var subHandle = client.onEvent('reconnect-successful',
+            var subHandle = client.onEvent('reconnect',
               function waitForConnectSuccess() {
-                eventCalled++;
+                resolve();
               });
-
-            setTimeout(function() {
-              expect(eventCalled).to.equal(1);
-              client.offEvent(subHandle);
-              resolve();
-            },35000)
           }
         });
       })
       .then(function () {
         return client.disconnect();
       });
+  });
+
+  it('does not retry if we are connected', function () {
+    var client;
+
+    return Happn.client.create(
+      {
+        config: {
+          username: testUser2.username,
+          password: testUser2.password
+        }
+      })
+      .then(function (clientInstance) {
+        client = clientInstance;
+        return stopService();
+      })
+      .then(function waitForDisconnect() {
+        return new Promise(function (resolve) {
+          {
+            var subHandle = client.onEvent('reconnect-scheduled',
+              function waitForReconnectScheduled() {
+                client.offEvent(subHandle);
+                resolve();
+              });
+          }
+        });
+      })
+      .then(function startServiceWithNoLogin() {
+        return createService(false);
+      })
+      .then(function waitForLoginAttempt() {
+        return new Promise(function (resolve) {
+          server1.services.pubsub.on('loginAttempt', function waitForAttempt() {
+            resolve();
+          })
+        });
+      })
+      .then(stopService)
+      .then(function () {
+        return createService(true);
+      })
+      .then(function () {
+        return new Promise(function (resolve) {
+          {
+            var subHandle = client.onEvent('reconnect-successful',
+              function waitForConnectSuccess() {
+                client.offEvent(subHandle);
+                resolve();
+              });
+          }
+        });
+      })
+      .then(function () {
+        return new Promise(function (resolve, reject) {
+          {
+            var subHandle = client.onEvent('reconnect',
+              function waitForConnectSuccess() {
+                reject(new Error('This should not happen'));
+              });
+            setTimeout(function () {
+              client.offEvent(subHandle);
+              resolve();
+            }, 20000);
+          }
+        });
+      })
+      .then(function () {
+        return client.disconnect();
+      });
+  });
+
+  it('will not reconnect after shutdown', function () {
+    var client;
+
+    return Happn.client.create(
+      {
+        config: {
+          username: testUser2.username,
+          password: testUser2.password
+        }
+      })
+      .then(function (clientInstance) {
+        client = clientInstance;
+        return stopService();
+      })
+      .then(function () {
+        return client.disconnect();
+      })
+      .then(function startServiceWithNoLogin() {
+        return createService(false);
+      })
+      .then(function () {
+        return new Promise(function (resolve) {
+          {
+            var eventCalled = 0;
+            var subHandle = client.onEvent('reconnect',
+              function waitForReconnect() {
+                eventCalled++;
+              });
+
+            setTimeout(function () {
+              expect(eventCalled).to.equal(0);
+              client.offEvent(subHandle);
+              resolve();
+            }, 20000)
+          }
+        });
+      })
   });
 });

